Make `DataPHA.get_x()` work if channels are not counted starting at 1.

hamogu opened the pull request:
## Summary
Make `DataPHA.get_x()` work if channels are not counted starting at 1.

## Details
See #1305 for details. OGIP allows any number as starting number to count channels, but calls out 0 and 1 as examples. Of course, this may break in other parts of the code, but let's fix them as we find them.

### Note
I'm pushing this up without running the tests locally because it's a pretty simple change and I don't have right now time to debug why I can't build sherpa locally any more. (some libraries not found).

DougBurke commented:
For PRs like this I've been trying to get into the habit [*] of

a) add tests
b) change code

where the tests in a may either be marked as xfail (for the cases where we can say 'this is obviously wrong') or just document the current behavior.

It makes it a bit easier to see what the changes in b do to the tests.

[*] which normally requires excessive use of rebasing to split apart commits and moving them around.

DougBurke commented:
So we have

- channel starting at 1 (I'd like to say that all issues are fixed but we have quality handling and probably some edge cases still hiding around)
- channel starting at 0
- channel starting at something other than 1 or 0 (I'd put this as less important but it should hppefully come along for free when handling the 0 case, as long as we  add tests)

and then there's PHA files which only contain a subset of the valid set of channels (which is a related, but separate, issue, in part because I'd put it at a lower priority than the channel=0 case).

I would like to have a "test" PR that adds a bunch of tests - not just for get_x - and in particular some "end-to-end" tests (e.g. including grouping and filtering and fitting) so that we can see how much a change like this one fixes things. I did recently add some very-basic channel=0 tests but they aren't really sufficient (e.g. they didn't highlight this problem).

hamogu commented:
For the record: To avoid the excessive rebasing mentioned by @DougBurke above, I've added the test first, verified locally that it failed except when the channels start at 1 and then implemented the (obvious) fix.

While end-to-end test are nice, I was really working on something else and don't want to do full sherpa spelunking now. This is an obvious bug, I added a test, I fixed it and everything else is out of the scope of this PR. Note also, that the fix in in a code back that is "unusual", i.e. only when no arf or rmf are set - a case we only see in development work on other crazy circumstances, while the "end-to-end" tests that were suggested above should most importantly cover our real use cases for DAtaPHA: X-ray data with arf / rmf.

hamogu commented:
I realize now that I run into this before - this PR should close #1040.

DougBurke commented:
In #1040 I was trying to point out that we need to think about how we read in files like this (ie with TLMINn != 1 for whatever n is the channel column) as I think that there are issues in both the pyfits and crates backends. Which is part of my "if we're going to fix this, let's try and actually make it workable" vibe from above. Last time I tried however I ended up being discouraged because of the changes we'd need to make to DataPHA (I think, I forget now).

hamogu commented:
I suggest to do this fairly isolated change because it goes in the right way - and it only effects DataPHA objects created "by-hand", which is different from reading anything from a file with either backend.

hamogu commented:
@DougBurke: I know it's been a while. Can I ask you to have a look again (the failure is only codecov, not something that matters here)?
I totally admit that it does not solve all problems we could possibly have with TLMIN != 1, but it is a very isolated change that goes in the right direction. At some point, we do want to be able to work with all TLMIN more broadly, but it's a mess. This change is only for a single code path - a DataPHA created by hand in init, not read in from a file. That's something we barely do, so there is essentially no risk of side effects or confusing users. It's clearly just a bug fix.

And later, when we get to solve all TLMIN problems, then we will be happy that we have this one in already. On the other hand, we we close this now, then we'll forget about it and will painfully rediscover it later. I discovered it twice already (in #1040 and in #1305), I don't want to discover it again.

DougBurke commented:
@hamogu - what is the underlying cause that makes you keep on hitting this? Is it that in your code you have `channel=np.arange(4)` - i.e. a channel range starting from 0 - or are you reading things in some other way. 

Also, why (for #1305) do you want a `DataPHA` object if you don't have a response (is it to get the grouping or some other reason)?

hamogu commented:
I want DataPHA in my example because it's a PHA set and I want to group it. If we had a different "DataBinned" class I could use that instead. Also, some of the plotting works without an RMF, I get the nice display of the data object in the notebook (listing the stuff that's relevant for PHA). 

Why do I keep hitting this? Because I work on calibration and instrument development. I've hit this in ARCUS work and in work for the Chandra ACIS gain calibration - both cases where we don't have a complete OGIP compliant set of PHA, ARF, RMF. In Arcus, I'm running the simulations to make the RMF, but in the process I need to fit PHA data. In Chandra calibration, we want to calibrate the GAIN, so the usual channels-to-energy conversion is not valid (I want to get that, after all) and this I can't (shouldn't) use the normal Chandra ARFs and RMFs. So, I do get `DataPHA` that I set up manually, e.g. I chop out something from the middle of a larger Chandra PHA file or I exctract (manually) something from just one Chandra or Arcus chip, and thus the channels range of the grating spectrum does not start at 0.

To repeat what I said above: 
> I totally admit that it does not solve all problems we could possibly have with TLMIN != 1, but it is a very isolated change that goes in the right direction.

I'm not quite sure I understand @DougBurke s reluctance to review this here. It's a bug, I implemented a test of the exiting behavior, verified that it failed, and then fixed the code.

hamogu commented:
\begin{rant}
@DougBurke asked me for "real-world examples". I assume that mean "example that use PHA files". Well, we don't have that because this code path is only used when `DataPHA` is created "by hand". If we don't want to support that, let's openly declare "Sherpa DataPHA objects can only be created from OGIP compliant files" and remove this code (and I'll write my own `MoritzDataPHA` class that add this functionality back in for me). But, as of now, the this code is here and it's incomplete, so what not fix it.
\end{rant}
